cvprac/cvp_api.py

```diff
@@ -3828,3 +3828,232 @@ def device_decommissioning_status_get_all(self, status="DECOMMISSIONING_STATUS_U
+        return self.clnt.post(url, data=payload)
+
+    def svc_account_account_get_one(self, username):
+        ''' Get an arbitrary service account's state using Resource APIs
+            Supported versions: CVP 2021.3.0 or newer and CVaaS.
+            Args:
+                username (string): The service account username.
+            Returns:
+                response (list): Returns a list of dict that contains...
+                Ex: [{'value': {'key': {'name': 'ansible'}, 'status': 'ACCOUNT_STATUS_ENABLED',
+                      'description': 'lab-tests', 'groups': {'values': ['network-admin']}},
+                      'time': '2022-02-10T04:28:14.251684869Z'}]
+        '''
+        payload = {"key":{"name": username}}
+        url = '/api/v3/services/arista.serviceaccount.v1.AccountConfigService/GetOne'
+        if not self.clnt.is_cvaas:
+            if self.clnt.apiversion is None:
+                self.get_cvp_info()
+            if self.clnt.apiversion < 7.0:
+                self.log.warning('Service Account Resource APIs are supported from 2021.3.0+')
+                return None
+        self.log.debug('v7 {} '.format(url) + str(payload))
+        return self.clnt.post(url, data=payload)
+
+    def svc_account_account_set(self, username, description, roles, status):
+        ''' Create a service account token using Resource APIs.
+            Supported versions: CVP 2021.3.0 or newer and CVaaS.
+            Args:
+                username (string): The service account username.
+                description (string): The description of the service account.
+                roles (list): The list of role IDs. Default roles have a human readable name,
+                    e.g.: 'network-admin', 'network-operator';
+                    other roles will have the format of 'role_<unix_timestamp>',
+                    e.g. 'role_1658850344592739349'.
+                    cvprac automatically converts non-default role names to role IDs.
+                status (enum): The status of the service account. Possible values:
+                    0 or 'ACCOUNT_STATUS_UNSPECIFIED'
+                    1 or 'ACCOUNT_STATUS_ENABLED'
+                    2 or 'ACCOUNT_STATUS_DISABLED'
+            Returns:
+                response (list): Returns a list of dict that contains...
+                Ex: [{'value': {'key': {'name': 'cvprac2'}, 'status': 'ACCOUNT_STATUS_ENABLED',
+                      'description': 'testapi', 'groups': {'values':
+                      ['network-admin', 'role_1658850344592739349']}},
+                      'time': '2022-07-26T18:19:55.392173445Z'}]
+        '''
+        role_ids = []
+        all_roles = self.get_roles()
+        for r in roles:
+            for role in all_roles['roles']:
+                if r == role['key']:
+                    role_ids.append(r)
+                elif r == role['name']:
+                    role_ids.append(role['key'])
```

Comment:
Is the issue here to look up the role ID for all the provided roles since it is not guaranteed that the role name is the same as the ID? Hence what we are really validating is that the roles provided to the function have their valid IDs put into the API payload?

Reply:
so the API only supports roleIDs, which is not great because besides the default roles (network-admin, network-operator) every other role has a role id in the format of `role_<unix timestamp>`, however users would have to find that role ID first and not many folks know how to do so, and it makes more sense to use names, so what I do above is even if the user provides the name, I swap it out with the ID 

so I'm getting all the roles, then loop through them, and check if the user input was a name or an ID, if it's an ID then I append it to a new list, if it's a name then I swap it out and append it to that new list, which I'll then use at the end to pass it in the API

Reply:
Thanks for clarification Tamas. Resolving original conversation as I made another suggestion later.